File: frontend/src/services/categorias.ts
import { http } from "./http"

// ===== Categorias de Despesas =====
export interface CategoriaDespesa {
  id: number
  nome: string
  tipo: "fixa" | "variavel"
  essencial: number
  cor: string | null
  icone: string | null
  ativo: number
  criado_em: string
  atualizado_em: string
}

export const catDespService = {
  listar: (apenasAtivos = false) =>
    http.get<CategoriaDespesa[]>("/categorias/despesas",
      { params: { apenas_ativos: apenasAtivos } }).then(r => r.data),
  criar: (p: Partial<CategoriaDespesa>) =>
    http.post<CategoriaDespesa>("/categorias/despesas", p).then(r => r.data),
  atualizar: (id: number, p: Partial<CategoriaDespesa>) =>
    http.put<CategoriaDespesa>(`/categorias/despesas/${id}`, p).then(r => r.data),
  inativar: (id: number) => http.delete(`/categorias/despesas/${id}`),
}

// ===== Categorias de Receitas =====
export interface CategoriaReceita {
  id: number
  nome: string
  recorrencia: "recorrente" | "eventual"
  cor: string | null
  ativo: number
  criado_em: string
  atualizado_em: string
}

export const catRecService = {
  listar: (apenasAtivos = false) =>
    http.get<CategoriaReceita[]>("/categorias/receitas",
      { params: { apenas_ativos: apenasAtivos } }).then(r => r.data),
  criar: (p: Partial<CategoriaReceita>) =>
    http.post<CategoriaReceita>("/categorias/receitas", p).then(r => r.data),
  atualizar: (id: number, p: Partial<CategoriaReceita>) =>
    http.put<CategoriaReceita>(`/categorias/receitas/${id}`, p).then(r => r.data),
  inativar: (id: number) => http.delete(`/categorias/receitas/${id}`),
}

File: backend/app/schemas/categorias.py
from pydantic import Field
from app.schemas.common import ORMBase, TimestampSchema


# ===== Despesas =====
class CategoriaDespesaBase(ORMBase):
    nome: str = Field(min_length=1, max_length=60)
    tipo: str = Field(default="variavel", pattern="^(fixa|variavel)$")
    essencial: int = Field(default=0, ge=0, le=1)
    cor: str | None = Field(default="#888888", max_length=7)
    icone: str | None = None


class CategoriaDespesaCreate(CategoriaDespesaBase):
    pass


class CategoriaDespesaUpdate(ORMBase):
    nome: str | None = None
    tipo: str | None = None
    essencial: int | None = None
    cor: str | None = None
    icone: str | None = None
    ativo: int | None = None


class CategoriaDespesaOut(CategoriaDespesaBase, TimestampSchema):
    id: int
    ativo: int


# ===== Receitas =====
class CategoriaReceitaBase(ORMBase):
    nome: str = Field(min_length=1, max_length=60)
    recorrencia: str = Field(default="eventual", pattern="^(recorrente|eventual)$")
    cor: str | None = "#28a745"


class CategoriaReceitaCreate(CategoriaReceitaBase):
    pass


class CategoriaReceitaUpdate(ORMBase):
    nome: str | None = None
    recorrencia: str | None = None
    cor: str | None = None
    ativo: int | None = None


class CategoriaReceitaOut(CategoriaReceitaBase, TimestampSchema):
    id: int
    ativo: int